perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,717 @@
+  std::shared_ptr<int> & ranks_feat_ptr, std::shared_ptr<int> & interval_starts_ptr,
+  std::shared_ptr<int> & interval_lengths_ptr)
+{
+  int num_points = n_img_ * depth_num_ * feat_h_ * feat_w_;
+  Eigen::Vector3f * frustum = new Eigen::Vector3f[num_points];
+
+  for (int i = 0; i < n_img_; i++) {
+    for (int d_ = 0; d_ < depth_num_; d_++) {
+      for (int h_ = 0; h_ < feat_h_; h_++) {
+        for (int w_ = 0; w_ < feat_w_; w_++) {
+          int offset =
+            i * depth_num_ * feat_h_ * feat_w_ + d_ * feat_h_ * feat_w_ + h_ * feat_w_ + w_;
+          (frustum + offset)->x() = static_cast<float>(w_) * (input_img_w_ - 1) / (feat_w_ - 1);
+          (frustum + offset)->y() = static_cast<float>(h_) * (input_img_h_ - 1) / (feat_h_ - 1);
+          (frustum + offset)->z() = static_cast<float>(d_) * depth_step_ + depth_start_;
+
+          // eliminate post transformation
+          *(frustum + offset) -= post_trans_.translation();
+          *(frustum + offset) = post_rot_.inverse() * *(frustum + offset);
+          //
+          (frustum + offset)->x() *= (frustum + offset)->z();
+          (frustum + offset)->y() *= (frustum + offset)->z();
+          // img to ego -> rot -> trans
+          *(frustum + offset) = cams2ego_rot_[i] * cams_intrin_[i].inverse() * *(frustum + offset) +
+                                cams2ego_trans_[i].translation();
+
+          // voxelization
+          *(frustum + offset) -= Eigen::Vector3f(x_start_, y_start_, z_start_);
+          (frustum + offset)->x() = static_cast<int>((frustum + offset)->x() / x_step_);
+          (frustum + offset)->y() = static_cast<int>((frustum + offset)->y() / y_step_);
+          (frustum + offset)->z() = static_cast<int>((frustum + offset)->z() / z_step_);
+        }
+      }
+    }
+  }
+
+  int * _ranks_depth = new int[num_points];
+  int * _ranks_feat = new int[num_points];
+
+  for (int i = 0; i < num_points; i++) {
+    _ranks_depth[i] = i;
+  }
+  for (int i = 0; i < n_img_; i++) {
+    for (int d_ = 0; d_ < depth_num_; d_++) {
+      for (int u = 0; u < feat_h_ * feat_w_; u++) {
+        int offset = i * (depth_num_ * feat_h_ * feat_w_) + d_ * (feat_h_ * feat_w_) + u;
+        _ranks_feat[offset] = i * feat_h_ * feat_w_ + u;
+      }
+    }
+  }
+
+  std::vector<int> kept;
+  for (int i = 0; i < num_points; i++) {
+    if (
+      static_cast<int>((frustum + i)->x()) >= 0 &&
+      static_cast<int>((frustum + i)->x()) < xgrid_num_ &&
+      static_cast<int>((frustum + i)->y()) >= 0 &&
+      static_cast<int>((frustum + i)->y()) < ygrid_num_ &&
+      static_cast<int>((frustum + i)->z()) >= 0 &&
+      static_cast<int>((frustum + i)->z()) < zgrid_num_) {
+      kept.push_back(i);
+    }
+  }
+
+  valid_feat_num_ = kept.size();
+  int * ranks_depth_host = new int[valid_feat_num_];
+  int * ranks_feat_host = new int[valid_feat_num_];
+  int * ranks_bev_host = new int[valid_feat_num_];
+  int * order = new int[valid_feat_num_];
```

Comment:
Memory Leak: This line allocates memory using `new` without a corresponding `delete[]`. Use `std::vector<int>` for better memory management.